File: 605_can_place_flowers.py
class Solution:
    def canPlaceFlowers(self, flowerbeds: list[int], n: int) -> bool:
        i = 0
        num_plantable = 0

        while i < len(flowerbeds):
            pre_clear = i == 0 or flowerbeds[i - 1] == 0
            clear = flowerbeds[i] == 0
            post_clear = i == len(flowerbeds) - 1 or flowerbeds[i + 1] == 0
            is_plantable = pre_clear and clear and post_clear

            if is_plantable:
                num_plantable += 1

            if not post_clear:
                i += 3
            elif not clear or is_plantable:
                i += 2
            else:
                i += 1

        return num_plantable >= n

File: 605-can-place-flowers.ts
function canPlaceFlowers(flowerbed: number[], n: number): boolean {
  let index = 0
  let numPlantable = 0
  while (index < flowerbed.length) {
    const preClear = index === 0 || flowerbed[index - 1] === 0
    const curClear = flowerbed[index] === 0
    const postClear =
      index === flowerbed.length - 1 || flowerbed[index + 1] === 0
    const isPlantable = preClear && curClear && postClear

    if (isPlantable) {
      numPlantable++
    }

    // Skip ahead logic based on several factors. Used for Optimization but also for
    // correctness (i.e. if we're consuming the current space make sure it's accounted
    // for for the next space).
    if (!postClear) {
      index += 3
    } else if (!curClear || isPlantable) {
      index += 2
    } else {
      index += 1
    }
  }

  return numPlantable >= n
}
